snapd-glib/snapd-notices-monitor.c

```diff
@@ -0,0 +1,166 @@
+
+static void begin_monitor(SnapdNoticesMonitor *self);
+
+static void monitor_cb(SnapdClient* source,
+                       GAsyncResult* res,
+                       SnapdNoticesMonitor *self) {
+
+    g_autoptr(GPtrArray) notices = NULL;
+    g_autoptr(GError) error = NULL;
+
+    notices = snapd_client_get_notices_finish(source, res, &error);
+
+    if (error != NULL) {
+        g_clear_object(&self->cancellable);
+        g_clear_pointer(&self->last_date_time, g_date_time_unref);
+        self->last_date_time_seconds = -1.0;
+        g_signal_emit_by_name(self, "error-event", error);
+        return;
+    }
+
+    if (g_cancellable_is_cancelled(self->cancellable)) {
+        g_clear_object(&self->cancellable);
+        g_clear_pointer(&self->last_date_time, g_date_time_unref);
+        self->last_date_time_seconds = -1.0;
+        return;
+    }
+
+    if ((error == NULL) && (notices != NULL)) {
+        for (int i=0; i < notices->len; i++) {
+            g_autoptr(SnapdNotice) notice = g_object_ref(notices->pdata[i]);
+            g_autoptr(GDateTime) last_occurred = (GDateTime *)snapd_notice_get_last_occurred(notice);
+            gdouble last_occurred_seconds = snapd_notice_get_last_occurred_seconds(notice);
+
+            if ((self->last_date_time == NULL) ||
+                (g_date_time_compare(self->last_date_time, last_occurred) == -1) ||
+                ((g_date_time_compare(self->last_date_time, last_occurred) == 0) && (self->last_date_time_seconds < last_occurred_seconds))) {
+                    g_clear_pointer (&self->last_date_time, g_date_time_unref);
+                    self->last_date_time = g_date_time_ref(last_occurred);
+                    self->last_date_time_seconds = snapd_notice_get_last_occurred_seconds(notice);
+                }
+            g_signal_emit_by_name(self, "notice-event", notice);
+        }
+    }
+    begin_monitor(self); // continue waiting for new notifications
+}
+
+static void begin_monitor(SnapdNoticesMonitor *self) {
+    snapd_client_set_notices_filter_by_date_seconds (self->client, self->last_date_time_seconds);
+    snapd_client_get_notices_async(self->client,
+                                   self->last_date_time,
+                                   200000000000000, // "infinity"
+                                   self->cancellable,
+                                   (GAsyncReadyCallback)monitor_cb,
+                                   self);
+}
+
+/**
+ * snapd_notices_monitor_start:
+ * @monitor: a #SnapdNoticesMonitor
+ * @cancellable: a #GCancellable
+ *
+ * Starts the asynchronous listening proccess, that will wait for new
+ * notices and emit a "notice-event" signal with the new notice as
+ * parameter.
+ *
+ * @return: TRUE if there was an error, FALSE if everything worked fine
+ * and the object is listening for events.
+ */
+gboolean snapd_notices_monitor_start(SnapdNoticesMonitor *self,
+                                 GCancellable *cancellable) {
```

Comment:
It's probably better to make the cancellable internal, and instead provide a `snapd_notices_monitor_stop` if this is required. The cancellable is important as it should be cancelled in dispose - this will stop the program crashing if SnapdNoticesMonitor is unreferenced.

Reply:
Mmm... good point!